File: src/normalization/financials.py
import logging
# math is imported inside but good to have
import math

def reconcile_financials(line_items: list, global_modifiers: dict, grand_total: float) -> list:
    """
    SMART DIRECTIONAL RECONCILIATION:
    Adjusts line items to match Grand Total based on mathematical directionality.
    """
    if not line_items or grand_total <= 0:
        return line_items
        
    current_sum = sum(float(item.get("Net_Line_Amount", 0)) for item in line_items)
    gap = current_sum - grand_total
    
    # Threshold for "Close Enough"
    # UPDATED: Tightened to 0.1% (or 0.5 Rs) because Pharma margins are thin.
    if abs(gap) < max(0.5, grand_total * 0.001):
        logger = logging.getLogger("normalization")
        logger.info(f"Reconcile: Sum {current_sum:.2f} matches Total {grand_total:.2f}. No changes.")
        return line_items

    logger = logging.getLogger("normalization")
    logger.info(f"Reconcile: GAP DETECTED. Sum {current_sum:.2f} vs Total {grand_total:.2f} (Gap {gap:.2f})")
    logger.info(f"Reconcile: Input Modifiers: {global_modifiers}")

    # Extract Modifiers
    g_disc = abs(float(global_modifiers.get("Global_Discount_Amount", 0) or 0))
    g_tax = abs(float(global_modifiers.get("Global_Tax_Amount", 0) or 0) + 
                float(global_modifiers.get("SGST_Amount", 0) or 0) + 
                float(global_modifiers.get("CGST_Amount", 0) or 0) + 
                float(global_modifiers.get("IGST_Amount", 0) or 0))
    freight = abs(float(global_modifiers.get("Freight_Charges", 0) or 0))
    
    modifier_to_apply = 0.0
    action = "NONE"
    
    if gap > 0:
        # Inflation. Reduce.
        logger.info(f"Reconcile: Inflation Detected (Gap +{gap:.2f}). Looking for Reducers...")
        if g_disc > 0:
            modifier_to_apply = -gap 
            action = "APPLY_DISCOUNT_CORRECTION"
            logger.info(f"Reconcile: Found Global Discount ({g_disc}). Applying Correction of -{gap:.2f} to match Total.")
        else:
            # Check for Implicit Discount (Small Gap < 5%)
            gap_percentage = gap / grand_total if grand_total > 0 else 0
            if gap_percentage < 0.05:
                modifier_to_apply = -gap
                action = "APPLY_IMPLICIT_DISCOUNT"
                logger.info(f"Reconcile: Implicit Discount Detected ({gap_percentage:.1%}). No explicit discount found, but gap is small. Force Reconciling.")
            else:
                logger.warning(f"Reconcile: No Discount found and gap ({gap_percentage:.1%}) is too large for implicit correction. Doing nothing (Safe Mode).")
            
    elif gap < 0:
        # Deflation. Increase.
        logger.info(f"Reconcile: Deflation Detected (Gap {gap:.2f}). Looking for Adders...")
        adder_sum = g_tax + freight
        if adder_sum > 0:
            modifier_to_apply = -gap # -(-gap) = +gap
            action = "APPLY_TAX_FREIGHT_CORRECTION"
            logger.info(f"Reconcile: Found Tax/Freight ({adder_sum}). Applying Correction of +{abs(gap):.2f} to match Total.")
        else:
             logger.warning("Reconcile: No Tax/Freight found to increase value. Doing nothing (Safe Mode).")

    if modifier_to_apply != 0:
        for item in line_items:
            original_net = float(item.get("Net_Line_Amount", 0))
            ratio = original_net / current_sum if current_sum > 0 else 0
            
            share = modifier_to_apply * ratio
            new_net = original_net + share 
            
            item["Net_Line_Amount"] = round(new_net, 2)
            
            qty = float(item.get("Standard_Quantity", 1) or 1)
            if qty > 0:
                item["Final_Unit_Cost"] = round(new_net / qty, 2)
            
            item["Logic_Note"] = item.get("Logic_Note", "") + f" [Reconcile: {action}]"

    return line_items

File: src/normalization/financials.js

import winston from 'winston';

const logger = winston.createLogger({
    level: 'info',
    format: winston.format.json(),
    transports: [
        new winston.transports.Console()
    ]
});

/**
 * SMART DIRECTIONAL RECONCILIATION:
 * Adjusts line items to match Grand Total based on mathematical directionality.
 * @param {Array} lineItems 
 * @param {Object} globalModifiers 
 * @param {number} grandTotal 
 * @returns {Array}
 */
export function reconcileFinancials(lineItems, globalModifiers, grandTotal) {
    if (!lineItems || lineItems.length === 0 || grandTotal <= 0) {
        return lineItems;
    }

    const currentSum = lineItems.reduce((sum, item) => sum + parseFloat(item.Net_Line_Amount || 0), 0);
    const gap = currentSum - grandTotal;

    // Threshold for "Close Enough": 0.1% or 0.5
    const threshold = Math.max(0.5, grandTotal * 0.001);

    if (Math.abs(gap) < threshold) {
        logger.info(`Reconcile: Sum ${currentSum.toFixed(2)} matches Total ${grandTotal.toFixed(2)}. No changes.`);
        return lineItems;
    }

    logger.info(`Reconcile: GAP DETECTED. Sum ${currentSum.toFixed(2)} vs Total ${grandTotal.toFixed(2)} (Gap ${gap.toFixed(2)})`);
    logger.info(`Reconcile: Input Modifiers: ${JSON.stringify(globalModifiers)}`);

    const gDisc = Math.abs(parseFloat(globalModifiers.Global_Discount_Amount || 0));

    // Sum taxes
    const gTax = Math.abs(
        parseFloat(globalModifiers.Global_Tax_Amount || 0) +
        parseFloat(globalModifiers.SGST_Amount || 0) +
        parseFloat(globalModifiers.CGST_Amount || 0) +
        parseFloat(globalModifiers.IGST_Amount || 0)
    );

    const freight = Math.abs(parseFloat(globalModifiers.Freight_Charges || 0));

    let modifierToApply = 0.0;
    let action = "NONE";

    if (gap > 0) {
        // Inflation. Reduce.
        logger.info(`Reconcile: Inflation Detected (Gap +${gap.toFixed(2)}). Looking for Reducers...`);
        if (gDisc > 0) {
            modifierToApply = -gap;
            action = "APPLY_DISCOUNT_CORRECTION";
            logger.info(`Reconcile: Found Global Discount (${gDisc}). Applying Correction of -${gap.toFixed(2)} to match Total.`);
        } else {
            const gapPercentage = grandTotal > 0 ? gap / grandTotal : 0;
            if (gapPercentage < 0.05) {
                modifierToApply = -gap;
                action = "APPLY_IMPLICIT_DISCOUNT";
                logger.info(`Reconcile: Implicit Discount Detected (${(gapPercentage * 100).toFixed(1)}%). No explicit discount found, but gap is small. Force Reconciling.`);
            } else {
                logger.warn(`Reconcile: No Discount found and gap (${(gapPercentage * 100).toFixed(1)}%) is too large for implicit correction. Doing nothing (Safe Mode).`);
            }
        }
    } else if (gap < 0) {
        // Deflation. Increase.
        logger.info(`Reconcile: Deflation Detected (Gap ${gap.toFixed(2)}). Looking for Adders...`);
        const adderSum = gTax + freight;
        if (adderSum > 0) {
            modifierToApply = -gap; // -(-gap) = +gap
            action = "APPLY_TAX_FREIGHT_CORRECTION";
            logger.info(`Reconcile: Found Tax/Freight (${adderSum}). Applying Correction of +${Math.abs(gap).toFixed(2)} to match Total.`);
        } else {
            logger.warn("Reconcile: No Tax/Freight found to increase value. Doing nothing (Safe Mode).");
        }
    }

    if (modifierToApply !== 0) {
        lineItems.forEach(item => {
            const originalNet = parseFloat(item.Net_Line_Amount || 0);
            const ratio = currentSum > 0 ? originalNet / currentSum : 0;
            const share = modifierToApply * ratio;
            const newNet = originalNet + share;

            item.Net_Line_Amount = parseFloat(newNet.toFixed(2));

            const qty = parseFloat(item.Standard_Quantity || 1);
            if (qty > 0) {
                item.Final_Unit_Cost = parseFloat((newNet / qty).toFixed(2));
            }

            item.Logic_Note = (item.Logic_Note || "") + ` [Reconcile: ${action}]`;
        });
    }

    return lineItems;
}
